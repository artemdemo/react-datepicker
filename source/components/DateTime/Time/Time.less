@import '../datetime-general';

.datetime-time {
    .datetime-table();
    width: initial;
    margin: 0 auto;
    text-align: center;
}

.datetime-time-input-wrap {
    width: 30px;
}

.datetime-time-input {
    width: 100%;
    box-sizing: border-box;
}
