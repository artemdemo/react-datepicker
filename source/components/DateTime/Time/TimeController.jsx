import React from 'react';
import PropTypes from 'prop-types';

import './TimeController.less';

class TimeController extends React.Component {
    handleUp() {
        const { onUp, timePart } = this.props;
        onUp(timePart);
    }

    handleDown() {
        const { onDown, timePart } = this.props;
        onDown(timePart);
    }

    render() {
        const { date, format } = this.props;
        return (
            <td>
                <div
                    className='datetime-time-controller
                               datetime-time-controller_up'
                    onClick={this.handleUp.bind(this)} />
                <div>{date.format(format)}</div>
                <div
                    className='datetime-time-controller
                               datetime-time-controller_down'
                    onClick={this.handleDown.bind(this)} />
            </td>
        );
    }
}

TimeController.propTypes = {
    date: PropTypes.shape({}).isRequired,
    format: PropTypes.string.isRequired,
    timePart: PropTypes.shape({}).isRequired,
    onUp: PropTypes.func.isRequired,
    onDown: PropTypes.func.isRequired,
};

export default TimeController;
