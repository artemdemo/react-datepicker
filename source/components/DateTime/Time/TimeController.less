@import '../styles/general';

.datetime-time-controller {
    cursor: pointer;
    padding: 13px;
    border-radius: 4px;
    position: relative;

    &:hover {
        background-color: fade(@hover-color, 50%);
    }

    &:after {
        position: absolute;
        font-weight: bold;
        width: 15px;
        height: 15px;
    }
}
    .datetime-time-controller_up {
        &::after {
            content: '‹';
            left: 30%;
            top: 25%;
            transform: rotate(90deg);
        }
    }
    .datetime-time-controller_down {
        &::after {
            content: '‹';
            left: 15%;
            top: 15%;
            transform: rotate(-90deg);
        }
    }
