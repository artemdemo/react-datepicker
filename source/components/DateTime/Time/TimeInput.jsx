import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

const TIME_REGEX = /^\d+$/;

class TimeInput extends React.Component {
    constructor(props) {
        super(props);

        const { value } = props;
        this.state = {
            value,
        };
    }

    handleChange(e) {
        const { onChange } = this.props;
        const newValue = e.target.value;
        if (newValue === '' || TIME_REGEX.test(newValue)) {
            this.setState({
                value: newValue,
            });
            onChange && onChange(newValue);
        }
    }

    render() {
        const { className } = this.props;
        const inputClass = classnames(className);
        return (
            <input
                type='text'
                className={inputClass}
                value={this.state.value}
                onChange={this.handleChange.bind(this)} />
        );
    }
}

TimeInput.displayName = 'TimeInput';

TimeInput.propTypes = {
    className: PropTypes.string,
    value: PropTypes.string,
    onChange: PropTypes.func,
};

TimeInput.defaultProps = {
    className: '',
    value: '',
    onChange: null,
};

export default TimeInput;
