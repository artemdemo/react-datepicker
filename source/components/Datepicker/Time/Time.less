@import '../datepicker-general';

.datepicker-time {
    .datepicker-table();
    width: initial;
    margin: 0 auto;
    text-align: center;
}

.datepicker-time-input-wrap {
    width: 30px;
}

.datepicker-time-input {
    width: 100%;
    box-sizing: border-box;
}
