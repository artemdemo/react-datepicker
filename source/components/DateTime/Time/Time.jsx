import React from 'react';
import PropTypes from 'prop-types';
import TimeController from './TimeController';

import './Time.less';
import TimeInput from './TimeInput';

const TIME_TYPES = {
    hour: 'hour',
    minute: 'minute',
    second: 'second',
    millisecond: 'millisecond',
    meridiem: 'meridiem',
};

const TIME_PARTS = [
    {
        type: TIME_TYPES.hour,
        regex: /h{1,2}|H{1,2}|k{1,2}/,
    },
    {
        type: TIME_TYPES.minute,
        regex: /m{1,2}/,
    },
    {
        type: TIME_TYPES.second,
        regex: /s{1,2}/,
    },
    {
        type: TIME_TYPES.millisecond,
        regex: /S{1,3}/,
    },
    {
        type: TIME_TYPES.meridiem,
        regex: /[aA]/,
    },
];

class Time extends React.Component {
    changeUp(part) {
        const { date, onChange } = this.props;
        if (part.type === TIME_TYPES.meridiem) {
            if (parseInt(date.format('H'), 10) > 11) {
                onChange(date.clone().subtract(12, TIME_TYPES.hour));
            } else {
                onChange(date.clone().add(12, TIME_TYPES.hour));
            }
        } else {
            onChange(date.clone().add(1, part.type));
        }
    }

    changeDown(part) {
        const { date, onChange } = this.props;
        if (part.type === TIME_TYPES.meridiem) {
            if (parseInt(date.format('H'), 10) > 11) {
                onChange(date.clone().subtract(12, TIME_TYPES.hour));
            } else {
                onChange(date.clone().add(12, TIME_TYPES.hour));
            }
        } else {
            onChange(date.clone().subtract(1, part.type));
        }
    }

    changeMilliseconds(milliseconds) {
        const { date, onChange } = this.props;
        onChange(date.clone().set(TIME_TYPES.millisecond, milliseconds));
    }

    renderController(part, index) {
        const { timeFormat, date } = this.props;
        const match = part.regex.exec(timeFormat);
        if (match) {
            if (part.type === TIME_TYPES.millisecond) {
                return (
                    <td>
                        <div className='datetime-time-input-wrap'>
                            <TimeInput
                                value={date.format('SSS')}
                                onChange={this.changeMilliseconds.bind(this)}
                                className='datetime-time-input' />
                        </div>
                    </td>
                );
            }
            return (
                <TimeController
                    key={`datetime-time-controller-${index}`}
                    format={match[0]}
                    timePart={part}
                    onUp={this.changeUp.bind(this)}
                    onDown={this.changeDown.bind(this)}
                    date={date} />
            );
        }
        return null;
    }


    render() {
        return (
            <table className='datetime-time'>
                <tbody>
                    <tr>
                        {TIME_PARTS.map((part, index) => {
                            const result = [];
                            const controller = this.renderController(part, index);
                            if (index !== 0 && controller) {
                                result.push(
                                    <td key={`datetime-time-controller-${index}-1`}>:</td>,
                                );
                            }
                            result.push(
                                controller,
                            );
                            return result;
                        })}
                    </tr>
                </tbody>
            </table>
        );
    }
}

Time.propTypes = {
    date: PropTypes.shape({}).isRequired,
    onChange: PropTypes.func.isRequired,
    timeFormat: PropTypes.string.isRequired,
};

export default Time;
