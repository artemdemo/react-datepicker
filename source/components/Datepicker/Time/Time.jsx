import React from 'react';
import PropTypes from 'prop-types';
import TimeController from './TimeController';

import './Time.less';

const TIME_TYPES = {
    hour: 'hour',
    minute: 'minute',
    second: 'second',
    millisecond: 'millisecond',
};

const TIME_PARTS = [
    {
        type: TIME_TYPES.hour,
        regex: /h{1,2}|H{1,2}|k{1,2}/,
    },
    {
        type: TIME_TYPES.minute,
        regex: /m{1,2}/,
    },
    {
        type: TIME_TYPES.second,
        regex: /s{1,2}/,
    },
    {
        type: TIME_TYPES.millisecond,
        regex: /S{1,3}/,
    },
];

class Time extends React.Component {
    changeUp(part) {
        const { date, onChange } = this.props;
        onChange(date.clone().add(1, part.type));
    }

    changeDown(part) {
        const { date, onChange } = this.props;
        onChange(date.clone().subtract(1, part.type));
    }

    renderController(part, index) {
        const { timeFormat, date } = this.props;
        const match = part.regex.exec(timeFormat);
        if (match) {
            return (
                <TimeController
                    key={`datepicker-time-controller-${index}`}
                    format={match[0]}
                    timePart={part}
                    onUp={this.changeUp.bind(this)}
                    onDown={this.changeDown.bind(this)}
                    date={date} />
            );
        }
        return null;
    }


    render() {
        return (
            <table className='datepicker-time'>
                <tbody>
                    <tr>
                        {TIME_PARTS.map((part, index) => {
                            const result = [];
                            const controller = this.renderController(part, index);
                            if (index !== 0 && controller) {
                                result.push(
                                    <td key={`datepicker-time-controller-${index}-1`}>:</td>,
                                );
                            }
                            result.push(
                                controller,
                            );
                            return result;
                        })}
                    </tr>
                </tbody>
            </table>
        );
    }
}

Time.propTypes = {
    date: PropTypes.shape({}).isRequired,
    onChange: PropTypes.func.isRequired,
    timeFormat: PropTypes.string.isRequired,
};

export default Time;
