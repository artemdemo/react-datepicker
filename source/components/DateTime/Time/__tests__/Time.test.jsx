import React from 'react';
import moment from 'moment';
import renderer from 'react-test-renderer';
import Time from '../Time';

describe('<Time />', () => {
    const date = moment(1502459094000);

    it('simple render, with time format "HH:mm:ss:SSS"', () => {
        const tree = renderer.create(
            <Time
                date={date}
                onChange={() => {}}
                timeFormat='HH:mm:ss:SSS'
            />,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('time format "HH:mm:ss"', () => {
        const tree = renderer.create(
            <Time
                date={date}
                onChange={() => {}}
                timeFormat='HH:mm:ss'
            />,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('time format "HH:mm"', () => {
        const tree = renderer.create(
            <Time
                date={date}
                onChange={() => {}}
                timeFormat='HH:mm'
            />,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
