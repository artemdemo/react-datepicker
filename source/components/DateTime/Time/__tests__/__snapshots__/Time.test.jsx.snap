// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Time /> simple render, with time format "HH:mm:ss:SSS" 1`] = `
<table
  className="datetime-time"
>
  <tbody>
    <tr>
      <td>
        <div
          className="datetime-time-controller datetime-time-controller_up"
          onClick={[Function]}
        />
        <div>
          16
        </div>
        <div
          className="datetime-time-controller datetime-time-controller_down"
          onClick={[Function]}
        />
      </td>
      <td>
        :
      </td>
      <td>
        <div
          className="datetime-time-controller datetime-time-controller_up"
          onClick={[Function]}
        />
        <div>
          44
        </div>
        <div
          className="datetime-time-controller datetime-time-controller_down"
          onClick={[Function]}
        />
      </td>
      <td>
        :
      </td>
      <td>
        <div
          className="datetime-time-controller datetime-time-controller_up"
          onClick={[Function]}
        />
        <div>
          54
        </div>
        <div
          className="datetime-time-controller datetime-time-controller_down"
          onClick={[Function]}
        />
      </td>
      <td>
        :
      </td>
      <td>
        <div
          className="datetime-time-input-wrap"
        >
          <input
            className="datetime-time-input"
            onChange={[Function]}
            type="text"
            value="000"
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Time /> time format "HH:mm" 1`] = `
<table
  className="datetime-time"
>
  <tbody>
    <tr>
      <td>
        <div
          className="datetime-time-controller datetime-time-controller_up"
          onClick={[Function]}
        />
        <div>
          16
        </div>
        <div
          className="datetime-time-controller datetime-time-controller_down"
          onClick={[Function]}
        />
      </td>
      <td>
        :
      </td>
      <td>
        <div
          className="datetime-time-controller datetime-time-controller_up"
          onClick={[Function]}
        />
        <div>
          44
        </div>
        <div
          className="datetime-time-controller datetime-time-controller_down"
          onClick={[Function]}
        />
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`<Time /> time format "HH:mm:ss" 1`] = `
<table
  className="datetime-time"
>
  <tbody>
    <tr>
      <td>
        <div
          className="datetime-time-controller datetime-time-controller_up"
          onClick={[Function]}
        />
        <div>
          16
        </div>
        <div
          className="datetime-time-controller datetime-time-controller_down"
          onClick={[Function]}
        />
      </td>
      <td>
        :
      </td>
      <td>
        <div
          className="datetime-time-controller datetime-time-controller_up"
          onClick={[Function]}
        />
        <div>
          44
        </div>
        <div
          className="datetime-time-controller datetime-time-controller_down"
          onClick={[Function]}
        />
      </td>
      <td>
        :
      </td>
      <td>
        <div
          className="datetime-time-controller datetime-time-controller_up"
          onClick={[Function]}
        />
        <div>
          54
        </div>
        <div
          className="datetime-time-controller datetime-time-controller_down"
          onClick={[Function]}
        />
      </td>
    </tr>
  </tbody>
</table>
`;
